File: src/main/java/io/disassemble/javanalysis/util/CodeParser.java
package io.disassemble.javanalysis.util;

import io.disassemble.javanalysis.CtMethodNode;
import io.disassemble.javanalysis.insn.*;
import javassist.bytecode.BadBytecode;
import javassist.bytecode.CodeAttribute;
import javassist.bytecode.CodeIterator;
import javassist.bytecode.ConstPool;

import java.util.ArrayList;
import java.util.List;

import static javassist.bytecode.Opcode.*;

/**
 * @author Tyler Sedlar
 * @since 5/20/2017
 */
public class CodeParser {

    public static List<CtInsn> parse(CtMethodNode method) {
        try {
            List<CtInsn> instructions = new ArrayList<>();
            ConstPool pool = method.pool();
            CodeAttribute attr = method.code();
            CodeIterator itr = attr.iterator();
            CtInsn previous = null;
            while (itr.hasNext()) {
                int index = itr.next();
                int opcode = itr.byteAt(index);
                int poolIndex = (index + 1);
                CtInsn insn;
                if (InsnUtil.isLdcInsn(opcode)) {
                    boolean wide = (opcode == LDC_W || opcode == LDC2_W);
                    poolIndex = (wide ? itr.u16bitAt(poolIndex) : itr.byteAt(poolIndex));
                    insn = new LdcInsn(method, index, opcode, poolIndex);
                } else if (InsnUtil.isIntInsn(opcode)) {
                    boolean s32 = (opcode == SIPUSH);
                    int operand = (s32 ? itr.s32bitAt(poolIndex) : itr.byteAt(poolIndex));
                    insn = new IntInsn(method, index, opcode, operand);
                } else if (InsnUtil.isVarInsn(opcode)) {
                    int var = (InsnUtil.isUnderscoreInsn(opcode) ? InsnUtil.underVal(opcode) : itr.byteAt(poolIndex));
                    insn = new VarInsn(method, index, opcode, var);
                } else if (InsnUtil.isJumpInsn(opcode)) {
                    boolean wide = (opcode == GOTO_W || opcode == JSR_W);
                    int target = ((wide ? itr.s32bitAt(poolIndex) : itr.s16bitAt(poolIndex)) + index);
                    insn = new JumpInsn(method, index, opcode, target);
                } else if (InsnUtil.isIncrementInsn(opcode)) {
                    int var = itr.byteAt(poolIndex);
                    int increment = itr.signedByteAt(poolIndex + 1);
                    insn = new IncrementInsn(method, index, opcode, var, increment);
                } else if (InsnUtil.isFieldInsn(opcode)) {
                    poolIndex = itr.u16bitAt(poolIndex);
                    String parent = pool.getFieldrefClassName(poolIndex);
                    String name = pool.getFieldrefName(poolIndex);
                    String desc = pool.getFieldrefType(poolIndex);
                    insn = new FieldInsn(method, index, opcode, parent, name, desc);
                } else if (InsnUtil.isMethodInsn(opcode)) {
                    poolIndex = itr.u16bitAt(poolIndex);
                    if (opcode == INVOKEINTERFACE) {
                        String parent = pool.getInterfaceMethodrefClassName(poolIndex);
                        String name = pool.getInterfaceMethodrefName(poolIndex);
                        String desc = pool.getInterfaceMethodrefType(poolIndex);
                        insn = new MethodInsn(method, index, opcode, parent, name, desc);
                    } else {
                        String parent = pool.getMethodrefClassName(poolIndex);
                        String name = pool.getMethodrefName(poolIndex);
                        String desc = pool.getMethodrefType(poolIndex);
                        insn = new MethodInsn(method, index, opcode, parent, name, desc);
                    }
                } else if (InsnUtil.isInvokeDynamicInsn(opcode)) {
                    poolIndex = itr.u16bitAt(poolIndex);
                    String type = pool.getInvokeDynamicType(poolIndex);
                    int nameIndex = pool.getInvokeDynamicNameAndType(poolIndex);
                    String name = pool.getUtf8Info(nameIndex); // TODO: check that this is correct.
                    int bootstrap = pool.getInvokeDynamicBootstrap(poolIndex);
                    insn = new InvokeDynamicInsn(method, index, opcode, name, type, bootstrap);
                } else if (InsnUtil.isTypeInsn(opcode)) {
                    poolIndex = itr.u16bitAt(poolIndex);
                    String type = pool.getClassInfo(poolIndex);
                    if (opcode == MULTIANEWARRAY) {
                        int dims = -1; // handle?
                        insn = new MultiANewArrayInsn(method, index, opcode, type, dims);
                    } else {
                        insn = new TypeInsn(method, index, opcode, type);
                    }
                } else if (opcode == WIDE) {
                    opcode = itr.byteAt(poolIndex);
                    index = itr.u16bitAt(poolIndex + 1);
                    insn = new CtInsn(method, index, opcode);
                } else if (InsnUtil.isTableSwitchInsn(opcode)) {
                    int key = (index & ~3) + 4;
                    int dfltIdx = (index + itr.s32bitAt(key));
                    int low = itr.s32bitAt(key += 4);
                    int high = itr.s32bitAt(key += 4);
                    int end = (high - low + 1) * 4 + (key += 4);
                    int size = low + ((end - index) / 4);
                    int[] indices = new int[size];
                    for (int idx = low; index < end; index += 4, idx++) {
                        indices[idx - low] = (index + itr.s32bitAt(key));
                    }
                    insn = new TableSwitchInsn(method, index, opcode, low, high, dfltIdx, indices);
                } else if (InsnUtil.isLookupSwitchInsn(opcode)) {
                    int key = (index & ~3) + 4;
                    int dfltIdx = (index + itr.s32bitAt(key));
                    int size = itr.s32bitAt(key += 4);
                    int[] keys = new int[size];
                    int[] indices = new int[size];
                    key += 4;
                    for (int i = 0; i < size; i++) {
                        int match = itr.s32bitAt(key);
                        int target = (index + itr.s32bitAt(key + 4));
                        keys[i] = match;
                        indices[i] = target;
                        key += 8;
                    }
                    insn = new LookupSwitchInsn(method, index, opcode, dfltIdx, keys, indices);
                } else {
                    insn = new CtInsn(method, index, opcode);
                }
                if (previous != null) {
                    previous.next.set(insn);
                    insn.previous.set(previous);
                }
                previous = insn;
                method.index(insn.index(), instructions.size());
                instructions.add(insn);
            }
            return instructions;
        } catch (BadBytecode e) {
            throw new IllegalStateException("Unable to parse instructions for " + method.source().getLongName());
        }
    }
}

File: src/main/java/io/disassemble/javanalysis/util/CodeParser.kt
package io.disassemble.javanalysis.util

import io.disassemble.javanalysis.code
import io.disassemble.javanalysis.insn.*
import javassist.CtMethod
import javassist.bytecode.BadBytecode
import javassist.bytecode.CodeIterator
import javassist.bytecode.Mnemonic
import javassist.bytecode.Opcode


/**
 * A simple utility class for parsing the bytecode instructions of a method.
 *
 * @author Tyler Sedlar
 * @since 3/30/2018
 */
object CodeParser : Opcode {

    private val opcodes = Mnemonic.OPCODE

    /**
     * Parses the bytecode instructions of the given [CtMethod].
     *
     * @param method The [CtMethod] to parse.
     *
     * @return A list of [CtInsn] contained within the given [CtMethod].
     */
    fun parse(method: CtMethod): List<CtInsn> {
        val insns: MutableList<CtInsn> = ArrayList()
        val info = method.methodInfo2
        info.codeAttribute ?: return insns // early return if no code attribute

        val iterator = method.code.iterator()

        var prevInsn: CtInsn? = null
        while (iterator.hasNext()) {
            val pos: Int
            try {
                pos = iterator.next()
            } catch (e: BadBytecode) {
                throw RuntimeException(e)
            }
            val insn = parse(method, iterator, pos)
            if (prevInsn != null) {
                prevInsn.next = insn
                insn.previous = prevInsn
            }
            prevInsn = insn
            insns.add(insn)
        }
        return insns
    }

    /**
     * Gets a string representation of the bytecode instruction at the specified
     * position.
     *
     * @param method The [CtMethod] to parse from.
     * @param iter The [CtMethod]'s [CodeIterator].
     * @param pos The position of the [CtInsn] to parse.
     *
     * @return The [CtInsn] at the given position.
     */
    fun parse(method: CtMethod, iter: CodeIterator, pos: Int): CtInsn {
        val opcode = iter.byteAt(pos)

        require(!(opcode > opcodes.size || opcode < 0)) { "Invalid opcode, opcode: $opcode pos: $pos" }

        when (opcode) {
            Opcode.BIPUSH, Opcode.SIPUSH -> {
                return IntInsn(method, pos)
            }
            Opcode.LDC, Opcode.LDC_W, Opcode.LDC2_W -> {
                return LdcInsn(method, pos)
            }
            Opcode.ILOAD, Opcode.ILOAD_0, Opcode.ILOAD_1, Opcode.ILOAD_2, Opcode.ILOAD_3,
            Opcode.LLOAD, Opcode.LLOAD_0, Opcode.LLOAD_1, Opcode.LLOAD_2, Opcode.LLOAD_3,
            Opcode.FLOAD, Opcode.FLOAD_0, Opcode.FLOAD_1, Opcode.FLOAD_2, Opcode.FLOAD_3,
            Opcode.DLOAD, Opcode.DLOAD_0, Opcode.DLOAD_1, Opcode.DLOAD_2, Opcode.DLOAD_3,
            Opcode.ALOAD, Opcode.ALOAD_0, Opcode.ALOAD_1, Opcode.ALOAD_2, Opcode.ALOAD_3,
            Opcode.ISTORE, Opcode.ISTORE_0, Opcode.ISTORE_1, Opcode.ISTORE_2, Opcode.ISTORE_3,
            Opcode.LSTORE, Opcode.LSTORE_0, Opcode.LSTORE_1, Opcode.LSTORE_2, Opcode.LSTORE_3,
            Opcode.FSTORE, Opcode.FSTORE_0, Opcode.FSTORE_1, Opcode.FSTORE_2, Opcode.FSTORE_3,
            Opcode.DSTORE, Opcode.DSTORE_0, Opcode.DSTORE_1, Opcode.DSTORE_2, Opcode.DSTORE_3,
            Opcode.ASTORE, Opcode.ASTORE_0, Opcode.ASTORE_1, Opcode.ASTORE_2, Opcode.ASTORE_3 -> {
                return VarInsn(method, pos)
            }
            Opcode.IFEQ, Opcode.IFGE, Opcode.IFGT, Opcode.IFLE, Opcode.IFLT, Opcode.IFNE,
            Opcode.IFNONNULL, Opcode.IFNULL, Opcode.IF_ACMPEQ, Opcode.IF_ACMPNE, Opcode.IF_ICMPEQ,
            Opcode.IF_ICMPGE, Opcode.IF_ICMPGT, Opcode.IF_ICMPLE, Opcode.IF_ICMPLT, Opcode.IF_ICMPNE,
            Opcode.GOTO, Opcode.JSR, Opcode.RET, Opcode.GOTO_W, Opcode.JSR_W -> {
                return JumpInsn(method, pos)
            }
            Opcode.IINC -> {
                return IncrementInsn(method, pos)
            }
            Opcode.TABLESWITCH -> {
                return TableSwitchInsn(method, pos)
            }
            Opcode.LOOKUPSWITCH -> {
                return LookupSwitchInsn(method, pos)
            }
            Opcode.GETSTATIC, Opcode.PUTSTATIC, Opcode.GETFIELD, Opcode.PUTFIELD -> {
                return FieldInsn(method, pos)
            }
            Opcode.INVOKEVIRTUAL, Opcode.INVOKESPECIAL, Opcode.INVOKESTATIC, Opcode.INVOKEINTERFACE -> {
                return MethodInsn(method, pos)
            }
            Opcode.INVOKEDYNAMIC -> {
                return InvokeDynamicInsn(method, pos)
            }
            Opcode.NEW, Opcode.NEWARRAY, Opcode.ANEWARRAY, Opcode.CHECKCAST -> {
                return TypeInsn(method, pos)
            }
            Opcode.MULTIANEWARRAY -> {
                return MultiANewArrayInsn(method, pos)
            }
            Opcode.WIDE -> {
                return WideInsn(method, pos)
            }
            else -> {
                return CtInsn(method, pos)
            }
        }
    }
}
